### PR TITLE
bluetooth: mesh: fix lpn receive window timing error

### PR DESCRIPTION
**Pull request by m-alperen-sener**
APIs for enabling and disabling the scanner may take larger time than expected and scheduling of the lpn state machine is affected by it. Since these type of latencies are factored into timeouts and scheduling times, we need to call scan enable/disable APIs after scheduling the next event time for LPN state machine to be more accurate.

### CODE REVIEW

**Review by alxelax — APPROVED** [review]
Nice catch. Did you try it with nrf53_ns?

**Comment by m-alperen-sener** [comment]
> Nice catch. Did you try it with nrf53_ns?

Yes

**Review by LingaoM — APPROVED** [review]

**Comment by alxelax** [comment]
Hi @PavelVPV, this requires your attention.

File: subsys/bluetooth/mesh/lpn.c [hunk 1]
```diff
@@ -939,8 +939,8 @@ static void lpn_timeout(struct k_work *work)
 	case BT_MESH_LPN_RECV_DELAY:
 		k_work_reschedule(&lpn->timer,
 				  K_MSEC(SCAN_LATENCY + lpn->recv_win + RX_DELAY_CORRECTION(lpn)));
-		bt_mesh_scan_enable();
 		lpn_set_state(BT_MESH_LPN_WAIT_UPDATE);
+		bt_mesh_scan_enable();
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/lpn.c:943]: This change has no effect as `lpn_set_state` doesn't do anything.

**Comment by PavelVPV** [comment]
> Since these type of latencies are factored into timeouts and scheduling times, we need to call scan enable/disable APIs after scheduling the next event time for LPN state machine to be more accurate.

This (`factored into timeouts`) is not mentioned for the `BT_MESH_LPN_RETRY_TIMEOUT` Kconfig option which controls the `FRIEND_REQ_RETRY_TIMEOUT` macro. With the PR change the Kconfig option description should say that the scanner enabling time should be considered in this timeout. **Also**, this option is measured in seconds, while the scanner enabling time can be measured in milliseconds. Iow, tuning this option doesn't allow an implementation to configure the rx radio time precisely.

We have the BT_MESH_LPN_SCAN_LATENCY` Kconfig option which is used to mitigate such issue. Perhaps we should use it in the rescheduling points you have changed in this PR?

Also, talking about `bt_mesh_scan_disable` relocation, I don't remember (and can't check at the moment) but guess that the scanner disabling time differs from the enabling time (and actually much shorter). Have you seen the difference in your measurements? My guess is that either the change for `bt_mesh_scan_disable` is not needed or a different Kconfig option should be used in this case.

**Review by PavelVPV — CHANGES_REQUESTED** [review]

**Comment by PavelVPV** [comment]
Hi @m-alperen-sener , do you happen to have screenshots from logic analyzer before and after this change?

**Comment by m-alperen-sener** [comment]
> > Since these type of latencies are factored into timeouts and scheduling times, we need to call scan enable/disable APIs after scheduling the next event time for LPN state machine to be more accurate.
> 
> This (`factored into timeouts`) is not mentioned for the `BT_MESH_LPN_RETRY_TIMEOUT` Kconfig option which controls the `FRIEND_REQ_RETRY_TIMEOUT` macro. With the PR change the Kconfig option description should say that the scanner enabling time should be considered in this timeout. **Also**, this option is measured in seconds, while the scanner enabling time can be measured in milliseconds. Iow, tuning this option doesn't allow an implementation to configure the rx radio time precisely.
> 
> We have the BT_MESH_LPN_SCAN_LATENCY` Kconfig option which is used to mitigate such issue. Perhaps we should use it in the rescheduling points you have changed in this PR?
> 
> Also, talking about `bt_mesh_scan_disable` relocation, I don't remember (and can't check at the moment) but guess that the scanner disabling time differs from the enabling time (and actually much shorter). Have you seen the difference in your measurements? My guess is that either the change for `bt_mesh_scan_disable` is not needed or a different Kconfig option should be used in this case.
 
The root cause of the problem: when we call the bt_mesh_scan_disable() within friend_response_received() the APIs calls take up to 6ms. Before the change, I observed that if the response is received at the end of the receive window, even though the next poll scheduled with correct time, the next poll starts almost immediately. What did I see is that the lpn_state is being changed to BT_MESH_LPN_ESTABLISHED but the actual call to lpn_timeout is because of this k_work schedule here https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L940.
 
When I measure the calls I saw that actual API calls are introducing delays, so I moved enable/disable calls after k_work schedule because we already schedule the k_work with that latency in mind. For scan enable; it is predictable and have a constant latency; but I can not say the same thing for scan_disable, the time it takes to return from API varies. However I don't think we should care much about them, because we always call scan disables when we already got what we wanted, either timeout happened or we actually received the response we waited for, the important part is immediately scheduling the k_work to next event accurately, then call disable scanner API which we can assume will disable the scanner ASAP, and whatever the call delay is, it will be still inside the k_work timeout time and it won't cause any time shift on the overall LPN state flow. Any response are received inside the extra window cause by scan disable latency will be or can be filtered out and rejected depending on the state.
 
> This change has no effect as `lpn_set_state` doesn't do anything.
 
For the above reasons it is better to keep state changes and k_work schedule next to each other and not introduce any delay in between, even just for conceptual reasons, although we know that in between delay will be zero or functionality will not differ at all.
 
What I wanted to imply with this `factored into timeouts` was only the latencies that we include in calculation, not the actual  friendship periodic timeouts.
 
before change

receive delay is about 10ms; receive window 17ms actual time the scanner is open ~15ms. Poll period 31ms. scan_disable API call took 4.2ms. When scan callback is called we spend all of the scan callback time for disabling the scanner, after the state is changed from BT_MESH_LPN_WAIT_UPDATE to BT_MESH_LPN_ESTABLISHED and next poll is scheduled. However during that time k_work for receive window timeout is already fired, before we reschedule it for next poll, thus the lpn_timeout k_work called but since the state is already changed to BT_MESH_LPN_ESTABLISHED, we send the next poll immediately.

after change one poll

after changes: receive window and scan window are the same however since we schedule the next period immediately in scan callback we are able to actually reschedule the k_work for the next poll.

after change poll periode

poll are periodic and triggered according to poll period.

**Comment by m-alperen-sener** [comment]
Logic analyzer capture records: don't forget to change file extensions to sal. Logic 2.
[with_change_sdc.zip](https://github.com/user-attachments/files/15535200/with_change_sdc.zip)
[with_change_zll.zip](https://github.com/user-attachments/files/15535201/with_change_zll.zip)
[without_change_sdc.zip](https://github.com/user-attachments/files/15535202/without_change_sdc.zip)
[without_change_zll.zip](https://github.com/user-attachments/files/15535203/without_change_zll.zip)

**Review by PavelVPV — APPROVED** [review]

**Comment by PavelVPV** [comment]
> > > Since these type of latencies are factored into timeouts and scheduling times, we need to call scan enable/disable APIs after scheduling the next event time for LPN state machine to be more accurate.
> > 
> > This (`factored into timeouts`) is not mentioned for the `BT_MESH_LPN_RETRY_TIMEOUT` Kconfig option which controls the `FRIEND_REQ_RETRY_TIMEOUT` macro. With the PR change the Kconfig option description should say that the scanner enabling time should be considered in this timeout. **Also**, this option is measured in seconds, while the scanner enabling time can be measured in milliseconds. Iow, tuning this option doesn't allow an implementation to configure the rx radio time precisely.
> > 
> > We have the BT_MESH_LPN_SCAN_LATENCY` Kconfig option which is used to mitigate such issue. Perhaps we should use it in the rescheduling points you have changed in this PR?
> > 
> > Also, talking about `bt_mesh_scan_disable` relocation, I don't remember (and can't check at the moment) but guess that the scanner disabling time differs from the enabling time (and actually much shorter). Have you seen the difference in your measurements? My guess is that either the change for `bt_mesh_scan_disable` is not needed or a different Kconfig option should be used in this case.
>  
> The root cause of the problem: when we call the bt_mesh_scan_disable() within friend_response_received() the APIs calls take up to 6ms. Before the change, I observed that if the response is received at the end of the receive window, even though the next poll scheduled with correct time, the next poll starts almost immediately. What did I see is that the lpn_state is being changed to BT_MESH_LPN_ESTABLISHED but the actual call to lpn_timeout is because of this k_work schedule here https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L940.
>  
> When I measure the calls I saw that actual API calls are introducing delays, so I moved enable/disable calls after k_work schedule because we already schedule the k_work with that latency in mind. For scan enable; it is predictable and have a constant latency; but I can not say the same thing for scan_disable, the time it takes to return from API varies. However I don't think we should care much about them, because we always call scan disables when we already got what we wanted, either timeout happened or we actually received the response we waited for, the important part is immediately scheduling the k_work to next event accurately, then call disable scanner API which we can assume will disable the scanner ASAP, and whatever the call delay is, it will be still inside the k_work timeout time and it won't cause any time shift on the overall LPN state flow. Any response are received inside the extra window cause by scan disable latency will be or can be filtered out and rejected depending on the state.
>  
> > This change has no effect as `lpn_set_state` doesn't do anything.
>  
> For the above reasons it is better to keep state changes and k_work schedule next to each other and not introduce any delay in between, even just for conceptual reasons, although we know that in between delay will be zero or functionality will not differ at all.
>  
> What I wanted to imply with this `factored into timeouts` was only the latencies that we include in calculation, not the actual  friendship periodic timeouts.
>  
> before change
> 
> receive delay is about 10ms; receive window 17ms actual time the scanner is open ~15ms. Poll period 31ms. scan_disable API call took 4.2ms. When scan callback is called we spend all of the scan callback time for disabling the scanner, after the state is changed from BT_MESH_LPN_WAIT_UPDATE to BT_MESH_LPN_ESTABLISHED and next poll is scheduled. However during that time k_work for receive window timeout is already fired, before we reschedule it for next poll, thus the lpn_timeout k_work called but since the state is already changed to BT_MESH_LPN_ESTABLISHED, we send the next poll immediately.
> 
> after change one poll
> 
> after changes: receive window and scan window are the same however since we schedule the next period immediately in scan callback we are able to actually reschedule the k_work for the next poll.
> 
> after change poll periode
> 
> poll are periodic and triggered according to poll period.
> 
> 
> 

Thanks for the detailed explanation!

**Pull request merged**